tools/sigma/backends/datadog.py

```diff
@@ -0,0 +1,103 @@
+# Output backends for sigmac
+# Copyright 2021 Datadog, Inc.
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import re
+
+from sigma.backends.base import SingleTextQueryBackend
+from sigma.parser.condition import NodeSubexpression
+
+
+class DatadogLogsBackend(SingleTextQueryBackend):
+    """Converts Sigma rule into Datadog log search query."""
+
+    identifier = "datadog-logs"
+    active = True
+    config_required = False
+
+    andToken = " AND "
+    orToken = " OR "
+    notToken = "-"
+    subExpression = "(%s)"
+    listExpression = "(%s)"
+    listSeparator = " OR "
+    valueExpression = "%s"
+    mapExpression = "%s:%s"
+    nullExpression = "-%s:*"
+    notNullExpression = "%s:*"
+
+    # The escaped characters list comes from https://docs.datadoghq.com/logs/explorer/search_syntax/#escaping-of-special-characters.
+    specialCharactersRegexp = re.compile(r'([+\-=&|><!(){}\[\]^"~?:\\/]+)')
+    whitespacesRegexp = re.compile(r"\s+")
+
+    facets = ["index", "service", "source"]
+
+    def __init__(self, sigmaconfig, backend_options=dict()):
+        if "index" in backend_options:
+            self.dd_index = backend_options["index"]
+
+        if "service" in backend_options:
+            self.dd_service = backend_options["service"]
+
+        if "source" in backend_options:
+            self.dd_source = backend_options["source"]
+
+        if sigmaconfig.config:
+            self.facets += sigmaconfig.config.get("facets", [])
+
+        super().__init__(sigmaconfig)
+
+    def generateQuery(self, parsed):
+        nodes = []
+
+        if hasattr(self, "dd_index"):
+            nodes.append(("index", self.dd_index))
+
+        if hasattr(self, "dd_service"):
+            nodes.append(("service", self.dd_service))
+
+        if hasattr(self, "dd_source"):
+            nodes.append(("source", self.dd_source))
+
+        if type(parsed.parsedSearch) == NodeSubexpression:
+            nodes.append(parsed.parsedSearch.items)
+        else:
+            nodes.append(parsed.parsedSearch)
+
+        return self.generateANDNode(nodes)
+
+    def cleanValue(self, val):
+        if type(val) == int:
+            return val
+        else:
+            return self.whitespacesRegexp.sub(
+                "?", self.specialCharactersRegexp.sub("\\\\\g<1>", val)
+            )
```

Comment:
Comments are appreciated to help better understand the regex.

For instance:
 - why add twice `\` in the final string?
 - why replace spaces with question marks?
   - `?` can match any char, not just space
   - `?` matches exactly one char, while we are replacing multiple contiguous spaces with a single `?`
   - why not `*` instead of `?`

Have we considered the alternative of quoting the value? This would solve both spaces and special characters escaping (at the cost of losing pattern matching)

```suggestion
            return '"' + val.replace('"', '\\"') + '"'
```

Reply:
These are good questions!

> why add twice `\` in the final string?

From [Python's regexp doc](https://docs.python.org/3/library/re.html):
> Regular expressions use the backslash character ('\') to indicate special forms or to allow special characters to be used without invoking their special meaning. This collides with Python’s usage of the same character for the same purpose in string literals; for example, to match a literal backslash, one might have to write '\\\\' as the pattern string, because the regular expression must be \\, and each backslash must be expressed as \\ inside a regular Python string literal.

This explains the unusual number of `\` in the `\\\\\g<1>` pattern.

> why replace spaces with question marks?

This is the only way to escape them and have the  `*` pattern matching work even in strings containing whitespaces.

>  `?` can match any char, not just space

Unfortunately, yes it can. We honestly don't see an ideal way of handling this, but we are opened to new ideas 😄 

>  `?` matches exactly one char, while we are replacing multiple contiguous spaces with a single `?`

Thanks for pointing this out. This was a bug. We improved the regexp in 6940bf4 to replace several whitespaces by the same number of `?`.

>  why not `*` instead of `?`

We don't see advantages in using `*`. It would give less guarantees than `?`s that at least can make sure we capture an exact number of characters and it would probably make queries slower since the search engine would try to resolve more matches.

> Have we considered the alternative of quoting the value? This would solve both spaces and special characters escaping (at the cost of losing pattern matching)

Yes and we decided not to use it in order to let the `*` pattern matching work even in strings containing whitespaces.

Reply:
This discussion is super interesting and definitely worth being recapped as a comment: this is the context that made you chose one implementation over another, and it's not something one can infer just by reading the code.

Reply:
Good call, I summarised it in a comment in 63705cd.